src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSource.cs

```diff
@@ -860,6 +891,31 @@ private void OnSuggestedActionsChanged(Workspace currentWorkspace, DocumentId cu
 
                 Volatile.Write(ref _lastSolutionVersionReported, solutionVersion);
             }
+
+            public async Task<ISuggestedActionCategorySet> GetSuggestedActionCategoriesAsync(ISuggestedActionCategorySet requestedActionCategories, SnapshotSpan range, CancellationToken cancellationToken)
+            {
+                var provider = _owner;
+                var result = new List<string>();
+                using (var asyncToken = _owner.OperationListener.BeginAsyncOperation(nameof(GetSuggestedActionCategoriesAsync)))
+                {
+                    var selection = await GetSpanAsync(range).ConfigureAwait(false);
+                    if (selection != null)
+                    {
+                        var document = range.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+                        var fixCategory = await GetFixLevelAsync(provider, document, range, cancellationToken).ConfigureAwait(false);
+                        if (fixCategory != null)
+                        {
+                            return _suggestedActionCategoryRegistry.CreateSuggestedActionCategorySet(fixCategory);
+                        }
+                        else if (await HasRefactoringsAsync(provider, document, selection, cancellationToken).ConfigureAwait(false))
```

Comment:
it feels like we could nicely parallelize things by running GetFixLevelAsync and HasRefactoringsAsync simultaneously.